src/util/arrayView.js

```diff
@@ -0,0 +1,74 @@
+// @flow
+
+import {type PluckFn} from "./sortBy";
+import sortBy from "./sortBy";
+
+export class ArrayView<T> {
+  _array: Array<T>;
+  _filteredArray: Array<T>;
+  _rowsPerPage: number;
+  _pages: number;
+
+  constructor(data: $ReadOnlyArray<T>) {
+    this._array = JSON.parse(JSON.stringify(data)); //Deep Copy
+    this._filteredArray = this._array;
+    this._rowsPerPage = this._array.length;
+    this._pages = 1;
+  }
+
+  /**
+    pluckArg: a function of an object that returns the element by which to sort
+    order: "ascending" or "descending"
+   */
+  sortBy(pluckArg: PluckFn<T>, order: "ascending" | "descending") {
+    this._array = sortBy(this._array, pluckArg);
+    this._filteredArray = sortBy(this._filteredArray, pluckArg);
+    if (order === "descending") {
+      this._array.reverse();
+      this._filteredArray.reverse();
+    }
+  }
+
+  /**
+    Removes any existing filters and applies the new filter if provided.
+
+    filterFn: a function that returns true for objects that should be included, omit to remove filter
+    return: the new number of pages
+   */
+  filter(filterFn?: (T) => boolean): number {
+    this._filteredArray = filterFn ? this._array.filter(filterFn) : this._array;
+    this._pages = Math.ceil(this._filteredArray.length / this._rowsPerPage);
+    return this._pages;
+  }
+
+  /**
+    rowsPerPage: how many rows should be displayed per page, +Infinity to display all rows on one page
+    return: the new number of pages
+   */
+  setRowsPerPage(rowsPerPage: number): number {
+    rowsPerPage = Math.min(rowsPerPage, this._filteredArray.length);
+    rowsPerPage = Math.max(rowsPerPage, 1);
+    rowsPerPage = parseInt(rowsPerPage.toFixed(0));
+    this._rowsPerPage = rowsPerPage;
+    this._pages = Math.ceil(this._filteredArray.length / this._rowsPerPage);
+    return this._pages;
+  }
+
+  /**
+    Get the total number of pages.
+   */
+  getPageCount(): number {
+    return this._pages;
+  }
+
+  /**
+    Get the contents of the specified page.
+   */
+  getPage(pageNumber: number): $ReadOnlyArray<T> {
+    pageNumber = parseInt(pageNumber.toFixed(0));
+    if (pageNumber < 1 || pageNumber > this._pages) return [];
```

Comment:
sure, will change to an error.

Reply:
Are you keeping the 1-based indexing? if so, why?

Reply:
Because the page number is ultimately a UI concept, which will be 1 indexed on the page selector front end element, it seems intuitive to me for `page` to be 1-indexed whereas `index` would be 0-indexed.